[STORM-1611] port org.apache.storm.pacemaker.pacemaker to java

hustfxj opened the pull request:
1 port pacemaker&pacemaker_test to java;
2 Update all the callings to cluster;
3 remove pacemaker.register about jmx;
4 fix the bug about #1071 , this is my fault on this worker.

hustfxj commented:
when we start up nimbus before starting up pacemaker, the nimbus will die. Because nimbus can't read the workers' heartbeats by "heartbeat-storms". In my opinion, it shouldn't be like this. @revans2  Do you think?

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,248 @@
+    private HBMessage sendPulse(HBPulse pulse) {
+        String id = pulse.get_id();
+        byte[] details = pulse.get_details();
+        LOG.debug("Saving Pulse for id [ {} ] data [ {} ].", id, details);
```
new String(details) will print the bytes correctly.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,248 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(Pacemaker.class);
+
+    private Map heartbeats;
```
this can be `private Map<String, byte[]> heartbeats`

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,248 @@
+    private HBMessage getPulse(String path, boolean authenticated) {
+        if (authenticated) {
+            byte[] details = (byte[]) heartbeats.get(path);
+            LOG.debug("Getting Pulse for path [ {} ]...data [ {} ].", path, details);
```
new String(details)

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,248 @@
+        while (true) {
+            int oldValue = pacemakerStats.averageHeartbeatSize.get();
+            int count = pacemakerStats.sendPulseCount.get();
+            newValue = ((count * oldValue) + newValue) / (count + 1);
```
This will keep modifying the newValue. I think it should be something like this
` newValue = ((count * oldValue) + size) / (count + 1);`

abhishekagarwal87 replied:
"value to add" should remain the same. Only the "value to set" should change.

hustfxj commented:
@abhishekagarwal87  Thank you. But I hope only it is the same as before for the details.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,246 @@
+            int oldValue = pacemakerStats.averageHeartbeatSize.get();
+            int count = pacemakerStats.sendPulseCount.get();
+            int newValue = ((count * oldValue) + size) / (count + 1);
+            if (!pacemakerStats.averageHeartbeatSize.compareAndSet(oldValue, newValue))
```
Minor. It can be simplified to 
```
if (pacemakerStats.averageHeartbeatSize.compareAndSet(oldValue, newValue))
 break;
```
Basically you can check for true condition and then break.

hustfxj commented:
Every can help me review this PR. In my perspective , I hope we can accelerate the first phrase job.

redsanket commented on storm-core/src/clj/org/apache/storm/daemon/nimbus.clj:
```diff
@@ -1353,8 +1353,8 @@
-                                        (StormCommon/topologyBases storm-cluster-state)))
+  (map-val #(clojurify-storm-base %) (clojurify-structure
+                                        (StormCommon/topologyBases storm-cluster-state))))
 
```
nice catch

redsanket commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,246 @@
+ */
+package org.apache.storm.pacemaker;
+
+import org.apache.storm.generated.*;
```
single class imports

redsanket commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,246 @@
+import org.slf4j.LoggerFactory;
+import uk.org.lidalia.sysoutslf4j.context.SysOutOverSLF4J;
+
+import java.util.*;
```
single class imports

redsanket commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,246 @@
+    private HBMessage exists(String path, boolean authenticated) {
+        HBMessage response = null;
+        if (authenticated) {
+            boolean itDoes = heartbeats.containsKey(path);
```
can it be renamed to pathExists?

redsanket commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,246 @@
+                int average = pacemakerStats.averageHeartbeatSize.getAndSet(0);
+                int totalKeys = heartbeats.size();
+                LOG.debug(
+                        "\nReceived {} heartbeats totaling {} bytes,\nSent {} heartbeats totaling {} bytes,\nThe largest heartbeat was {} bytes,\nThe average heartbeat was {} bytes,\nPacemaker contained {} total keys\nin the last {} second(s)",
```
can we divide this line into multiple lines for better readability

redsanket commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,246 @@
+                return sleepSeconds; // Run only once.
+            }
+        };
+        Utils.asyncLoop(afn, true, null, Thread.currentThread().getPriority(), false, true, null);
```
it will be nice to replace true, false with variables just to know what are being passed, but it is upto you

redsanket commented on storm-core/src/clj/org/apache/storm/pacemaker/pacemaker.clj:
```diff
@@ -1,242 +0,0 @@
-  (jmx/register-mbean
-    (jmx/create-bean
-      last-five-s)
-    "org.apache.storm.pacemaker.pacemaker:stats=Stats_Last_5_Seconds"))
```
Just curious where is the jmx related code implemented. Am I missing something or is it implemented else where

knusbaum replied:
`(:require [clojure.java.jmx :as jmx])`

redsanket commented:
Most of the changes look good, just a question about jmx reporter implementation

hustfxj commented:
@redsanket @abhishekagarwal87  I have addressed your comments, thank you. @redsanket  I also am curious where is the jmx related code implemented. But I don't found, so I removed pacemaker.register () about jmx.

abhishekagarwal87 commented:
LGTM

redsanket commented:
@hustfxj I guess it is more like a plugin to get metrics. I presume the code should be implemented. @knusbaum might be right to answer it.

knusbaum commented:
@hustfxj @redsanket 
The JMX stuff was in a clojure library `(:require [clojure.java.jmx :as jmx])` It should probably be moved over to Coda Hale since that's what all the other daemons are using now.

hustfxj commented:
@redsanket @knusbaum  thank you. I have addressed your comments

redsanket commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,372 @@
+import uk.org.lidalia.sysoutslf4j.context.SysOutOverSLF4J;
+
+
+import javax.management.*;
```
single imports

redsanket commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,372 @@
+        };
+        private PacemakerStats stats;
+
+        public PaceMakerDynamicMBean(PacemakerStats stats) {
```
minor nit "Pacemaker" instead of "PaceMaker"

knusbaum commented:
I'd prefer to go to Coda Hale for the stats like the rest of the daemons do. Take a look at DrpcServer.java for an example.

knusbaum commented on storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java:
```diff
@@ -0,0 +1,374 @@
+        }
+    }
+
+    public Pacemaker(Map conf, boolean isRegisterJmx) {
```
registering jmx isn't something we want in the constructor. I know stack traces bubble up, but I don't think anything bad actually happens from double registering. If we really want to avoid double-registering, we need a better mechanism than this, but I'm in favor of just letting double-registration occur.

hustfxj commented:
@knusbaum  I also prefer to go to Coda Hale for the stats before, It means we will drop the JMX Mbean. But I agree with you totaly.

hustfxj commented:
@knusbaum  I have remove the JMX, and use Gauge report the stats. I think Gauge is enough. can you look at it again?

knusbaum commented:
By using gauges we're re-implementing stuff that coda hale does, namely windowed stats. Just take a look at DRPCServer.java, like I mentioned. 

You can use Meters or Histograms (I don't care). But they should both allow you to delete a large amount of code. It should basically come down to 3-5 lines initializing the meters/histograms. Some of the current ones will be combined, for instance `largestHeartbeatSize` and `averageHeartbeatSize` can be combined into one Histogram automatically. Just set up a Histogram and mark it with the heartbeat size whenever a heartbeat comes in. 

Most of the work is just deleting code.

hustfxj commented:
@redsanket @knusbaum @revans2  can you look it again?

knusbaum commented:
+1

Metrics look great.

redsanket commented:
+1
